icu4c/source/i18n/number_longnames.cpp

```diff
@@ -344,38 +620,59 @@ class DerivedComponents {
         UnicodeString val1 = ures_getUnicodeStringByIndex(stackBundle.getAlias(), 1, &status);
         if (U_SUCCESS(status)) {
             if (val0.compare(UnicodeString(u"compound")) == 0) {
-                sp0 = compoundValue;
+                compound0_ = true;
             } else {
-                memory0.appendInvariantChars(val0, status);
-                sp0 = memory0.toStringPiece();
+                compound0_ = false;
+                value0_.appendInvariantChars(val0, status);
             }
             if (val1.compare(UnicodeString(u"compound")) == 0) {
-                sp1 = compoundValue;
+                compound1_ = true;
             } else {
-                memory1.appendInvariantChars(val1, status);
-                sp1 = memory1.toStringPiece();
+                compound1_ = false;
+                value1_.appendInvariantChars(val1, status);
             }
         }
     }
-    // The returned StringPiece is only valid as long as both the instance
-    // exists, and the compoundValue passed to the constructor is valid.
-    StringPiece value0() const {
-        return sp0;
+
+    // The returned StringPiece is only valid as long as the instance exists.
```

Comment:
optional:
`// Returns a StringPiece that is only valid.... `
or
`// NOTE: the returned .....`

Reply:
Done.